sklearn/_fast_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,749 @@
+        n_leaves = sum(estimator.get_n_leaf_nodes()
+                       for estimator in predictors_of_ith_iteration)
+
+        if n_trees == 1:
+            log_msg += ("{} tree, {} leaves, ".format(n_trees, n_leaves))
+        else:
+            log_msg += ("{} trees, {} leaves ".format(n_trees, n_leaves))
+            log_msg += ("({} on avg), ".format(int(n_leaves / n_trees)))
+
+        log_msg += "max depth = {}, ".format(max_depth)
+
+        if self.do_early_stopping_:
+            name = 'neg-loss' if self.scoring == 'loss' else 'score'
+            log_msg += "train {}: {:.5f}, ".format(name, self.train_score_[-1])
+            if self.validation_fraction is not None:
+                log_msg += "val {}: {:.5f}, ".format(
+                    name, self.validation_score_[-1])
+
+        iteration_time = time() - iteration_start_time
+        log_msg += "in {:0.3f}s".format(iteration_time)
+
+        print(log_msg)
+
+    def _raw_predict(self, X):
+        """Return the sum of the leaves values over all predictors.
+
+        Parameters
+        ----------
+        X : array-like, shape=(n_samples, n_features)
+            The input samples.
+
+        Returns
+        -------
+        raw_predictions : array, shape (n_samples * n_trees_per_iteration,)
+            The raw predicted values.
+        """
+        X = check_array(X, dtype=[X_DTYPE, X_BINNED_DTYPE])
+        check_is_fitted(self, 'estimators_')
+        if X.shape[1] != self.n_features_:
+            raise ValueError(
+                'X has {} features but this estimator was trained with '
+                '{} features.'.format(X.shape[1], self.n_features_)
+            )
+        is_binned = self._in_fit and X.dtype == X_BINNED_DTYPE
+        n_samples = X.shape[0]
+        raw_predictions = np.zeros(
+            shape=(self.n_trees_per_iteration_, n_samples),
+            dtype=self.baseline_prediction_.dtype
+        )
+        raw_predictions += self.baseline_prediction_
+        for predictors_of_ith_iteration in self.estimators_:
+            for k, estimator in enumerate(predictors_of_ith_iteration):
+                predict = (estimator.predict_binned if is_binned
+                           else estimator.predict)
+                raw_predictions[k, :] += predict(X)
+
+        return raw_predictions
+
+    @abstractmethod
+    def _get_loss(self):
+        pass
+
+    @abstractmethod
+    def _encode_y(self, y=None):
+        pass
+
+    @property
+    def n_estimators_(self):
+        check_is_fitted(self, 'estimators_')
+        return len(self.estimators_)
+
+
+class FastGradientBoostingRegressor(BaseFastGradientBoosting, RegressorMixin):
+    """Fast Gradient Boosting Regression Tree.
+
+    This estimator is much faster than
+    :class:`GradientBoostingRegressor<sklearn.ensemble.GradientBoostingRegressor>`
+    for big datasets (n_samples >= 10 000). The input data `X` is pre-binned
```

Comment:
Surely this pre-binning is something the user can do, so it's not a very compelling explanation of the considerable speed.

Reply:
It's true that users can bin the data themselves, but here since binning is always performed, this allows the algorithm to leverage interger-based data structures which make the training faster (in particular, the histograms and the histogram subtraction trick).

Binning alone (without taking parallelization into account) is responsible for the great majority of the performance boost.

As mentionned in my comment below I'm definitely planning on writing more details about this implementation in the user guide / docstrings, but if possible I'd like to do it in another PR.

Reply:
The speed difference is caused by the use of bin-histograms to find the splits (instead of sorting the features). And also we evaluate the split candidates only on a limited of fixed split levels (the bin boundaries).

This is not exactly the same model: working with bin-histograms can be considered as a regularized version of traditional decision trees.